pkg/manifest/manifest.go

```diff
@@ -132,15 +185,14 @@ func (m *Manifest) Include(excludeIdentifier *string, includeTechPreview *bool,
 		}
 	}
 
-	if includeTechPreview != nil {
-		featureGateAnnotation := "release.openshift.io/feature-gate"
-		featureGateAnnotationValue, featureGateAnnotationExists := annotations[featureGateAnnotation]
-		if featureGateAnnotationValue == string(configv1.TechPreviewNoUpgrade) && !(*includeTechPreview) {
-			return fmt.Errorf("tech-preview excluded, and %s=%s", featureGateAnnotation, featureGateAnnotationValue)
+	if requiredFeatureSet != nil {
+		shouldInclude, err := ResourceSatisfiesFeatureSetRequirement(*requiredFeatureSet, annotations)
+		if err != nil {
+			return err
 		}
-		// never include the manifest if the feature-gate annotation is outside of allowed values (only TechPreviewNoUpgrade is currently allowed)
-		if featureGateAnnotationExists && featureGateAnnotationValue != string(configv1.TechPreviewNoUpgrade) {
-			return fmt.Errorf("unrecognized value %s=%s", featureGateAnnotation, featureGateAnnotationValue)
+		if !shouldInclude {
+			manifestFeatureSets, _, _ := GetFeatureSetsFromAnnotations(annotations) // best effort for message
+			return fmt.Errorf("%q is required, and %s=%s", *requiredFeatureSet, featureSetAnnotation, strings.Join(manifestFeatureSets.List(), ","))
```

Comment:
you shouldn't need to handle this error message in `Include`.  If `ResourceSatisfiesFeatureSetRequirement` (or [a successor method][1]) doesn't think the manifest is suitable, it can construct the error, and `Include` can blindly bubble that error up.  Basically, no need to distinguish `shouldInclude, err`; you can just return `err`, with all the different categories of "we weren't sure we could include this" lumped into `err` and passed up the chain.

[1]: https://github.com/openshift/library-go/pull/1389#discussion_r950599219